posthog/async_migrations/migrations/0001_events_sample_by.py

```diff
@@ -0,0 +1,139 @@
+from constance import config
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.events import EVENTS_TABLE
+from posthog.async_migrations.definition import AsyncMigrationDefinition, AsyncMigrationOperation
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+from posthog.version_requirement import ServiceVersionRequirement
+
+TEMPORARY_TABLE_NAME = "temp_events_0001_events_sample_by"
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = "Events table migration for compatible sample by column."
+
+    posthog_min_version = "1.30.0"
+    posthog_max_version = "1.31.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
+    ]
+
+    operations = [
+        AsyncMigrationOperation(
```

Comment:
It would be lovely to have a summary here. It looks like we're doing two things here:

 1. use cityHash64 for event table sample by, [as required](https://clickhouse.com/docs/en/sql-reference/statements/alter/sample-by/)
 2. updating the `order by` on events table to use cityHash64 on distinct_id, and uuid. (Is it required to hash the uuid? I can understand the distinct_id as we'd want to order by the sampling key. I'm not sure what the ramifications are tbh. I would be curious about how this change to use hash as the order affects for queries where we want to get the first by distinct id for instance. But I haven't got a good handle about how this all works so add a pinch of salt)

For the migration steps it looks like we are:

 1. creating a new table with updated settings
 2. insert events from before last month from old to new table. I presume this is to avoid fighting new events coming in? Is that required?
 3. disable materialized view. Not sure what this materialized view is used for, what is the impact of detatching?
 4. Fill in the rest of the new events table with the remaining events in the old table. This assumes that there will not have been any events with timestamps with less than last month written to the old events table. Is this true? .e.g can plugins end up writing older events?
 5. rename the old table to some place safe, and rename new events table to `events`. This operation isn't atomic, so there will be a period where the app will get table not found I presume. Might be fine as it's a short amount of time but worth mentioning. This does also mean the rollback might not be successful.
 6. reattach materialized view
 7. make sure new events table is fully merged. Is there a performance hit whilst this is happening?

Not well informed understanding, but would be handy to have something like this, with the potential issues highlighted and accepted for instance.

Reply:
Cool, I'll write a docstring

Reply:
Also, is this an offline migration btw?

Reply:
Nope. The whole point of having built this whole system is that these migrations can happen in the background of a running instance.

Although it does cause partial disruption during the period when the materialized view is detached, as no events are ingested then.